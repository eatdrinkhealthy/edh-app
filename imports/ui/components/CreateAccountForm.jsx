// @flow
import React from "react";
import { Field, reduxForm } from "redux-form";
import { validateUserField } from "../../api/accounts/userSchema";
import FormInput from "./FormInput";
import { Button } from "./ReactBootstrapLib";

const validateUsername = value => validateUserField("username", value);

const validateEmail = value => validateUserField("email", value);

const validatePassword = value => validateUserField("password", value);

const validateConfirmPassword = (value, allValues) => (
  allValues.password !== allValues.confirmPassword
    ? "Password and Confirm Password fields do not match."
    : undefined
);

export type ICreateAccountFormValues = {
  username: string,
  email: string,
  password: string,
  confirmPassword: string,
};

type ICreateAccountFormProps = {
  handleSubmit: (values: ICreateAccountFormValues) => void,
  invalid: boolean,
};

// NOTE: when we call the CreateAccountForm component, we pass our submit handler as an
// onSubmit prop, and redux-form passes a submit function in as handleSubmit
const CreateAccountFormComponent = ({ handleSubmit, invalid }: ICreateAccountFormProps) => (
  <form className="measure center" onSubmit={handleSubmit}>
    <Field
      inputId="username"
      name="username"
      type="text"
      label="Username"
      component={FormInput}
      validate={validateUsername}
      autoFocus
    />
    <Field
      inputId="email"
      name="email"
      type="email"
      label="Email"
      component={FormInput}
      validate={validateEmail}
    />
    <Field
      inputId="password"
      name="password"
      type="password"
      label="Password"
      component={FormInput}
      validate={validatePassword}
    />
    <Field
      inputId="confirmPassword"
      name="confirmPassword"
      type="password"
      label="Confirm Password"
      component={FormInput}
      validate={validateConfirmPassword}
    />
    <Button id="createAccountSubmit" type="submit" disabled={invalid}>Create Account</Button>
  </form>
);

const onSubmitSuccess = (result, dispatch, props) => {
  // on successful submit, clear password fields
  props.change("password", "");
  props.untouch("password");          // prevents showing validation error
  props.change("confirmPassword", "");

  // set focus on username field
  // NOTE: using document.querySelector is not a React standard (works here, not in jest)
  const usernameElem = document.querySelector("input#username");
  if (usernameElem) {
    usernameElem.focus();
  }
};

const CreateAccountForm = reduxForm({
  form: "CreateAccountForm",
  onSubmitSuccess,
})(CreateAccountFormComponent);

export default CreateAccountForm;
