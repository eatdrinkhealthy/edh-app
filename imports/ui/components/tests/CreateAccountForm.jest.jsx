// @flow
/* eslint-env jest */
/* eslint-disable func-names, prefer-arrow-callback, no-unused-expressions */
/* eslint-disable import/no-extraneous-dependencies */
import React from "react";
import { createStore } from "redux";
import { Provider } from "react-redux";
import renderer from "react-test-renderer";
import toJson from "enzyme-to-json";
import CreateAccountForm from "../CreateAccountForm";
import mountFormWithInputs from "../../../utils/tests/mountFormWithInputs";
import appReducer from "../../../state/reducers";

describe("<CreateAccountForm />", function () {
  const testStore = createStore(appReducer);

  it("matches render snapshot - with form validation errors", function () {
    // render form with no input values to display validation errors
    const tree = renderer.create(
      <Provider store={testStore}>
        <CreateAccountForm onSubmit={() => {}} />
      </Provider>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("matches render snapshot - without form validation errors", function () {
    const props = {
      onSubmit: jest.fn(),
    };

    // mount form with valid input values to suppress form validation errors
    const wrapper = mountFormWithInputs(
      <CreateAccountForm {...props} />,
      {
        username: "user12",
        email: "user12@test.com",
        password: "user12pw",
        confirmPassword: "User12pw2",
      },
      testStore,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should set confirmPassword error and NOT call onSubmit, when password !== confirm password", function () {
    const props = {
      onSubmit: jest.fn(),
    };

    const wrapper = mountFormWithInputs(
      <CreateAccountForm {...props} />,
      {
        username: "user12",
        email: "user12@test.com",
        password: "user12pw",
        confirmPassword: "User12pw2",
      },
      testStore,
    );

    wrapper.find("input[type='submit']").simulate("submit");
    expect(wrapper.find("#confirmPasswordError").text())
      .toBe("Password and Confirm Password fields do not match.");
    expect(props.onSubmit).not.toHaveBeenCalled();
  });

  it("should set username too short error, when username less than min length", function () {
    const props = {
      onSubmit: jest.fn(),
    };

    const wrapper = mountFormWithInputs(
      <CreateAccountForm {...props} />,
      {
        username: "abc",
        email: "user12@test.com",
        password: "user12pw",
        confirmPassword: "User12pw2",
      },
      testStore,
    );

    expect(wrapper.find("#usernameError").text())
      .toBe("Username must be at least 4 characters.");
  });

  it("should set invalid email error, when email format is invalid", function () {
    const props = {
      onSubmit: jest.fn(),
    };

    const wrapper = mountFormWithInputs(
      <CreateAccountForm {...props} />,
      {
        username: "abcd",
        email: "user12",
        password: "user12pw",
        confirmPassword: "User12pw2",
      },
      testStore,
    );

    expect(wrapper.find("#emailError").text())
      .toBe("Email address must be a valid email address format.");
  });

  it("should call onSubmit, when submitting and all form fields are valid", function () {
    const props = {
      onSubmit: jest.fn(),
    };

    const wrapper = mountFormWithInputs(
      <CreateAccountForm {...props} />,
      {
        username: "user12",
        email: "user12@test.com",
        password: "user12pw",
        confirmPassword: "user12pw",
      },
      testStore,
    );

    wrapper.find("input[type='submit']").simulate("submit");
    expect(props.onSubmit).toHaveBeenCalledWith(
      {
        username: "user12",
        email: "user12@test.com",
        password: "user12pw",
        confirmPassword: "user12pw",
      },
      expect.anything(),
      expect.anything(),
    );
  });

  it("should clear password and confirmPassword on successful submit", function () {
    const props = {
      onSubmit: jest.fn(),
    };

    const wrapper = mountFormWithInputs(
      <CreateAccountForm {...props} />,
      {
        username: "user12",
        email: "user12@test.com",
        password: "user12pw",
        confirmPassword: "user12pw",
      },
      testStore,
    );
    const passwordNode = wrapper.find("input#password");
    const confirmPasswordNode = wrapper.find("input#confirmPassword");

    wrapper.find("input[type='submit']").simulate("submit");

    expect(passwordNode.props().value).toEqual("");
    expect(confirmPasswordNode.props().value).toEqual("");
  });

  it("should set focus to username input on render", function () {
    const props = {
      onSubmit: jest.fn(),
    };

    const wrapper = mountFormWithInputs(
      <CreateAccountForm {...props} />,
      {
        username: "user12",
        email: "user12@test.com",
        password: "user12pw",
        confirmPassword: "user12pw",
      },
      testStore,
    );

    expect(wrapper.find("input#username").get(0)).toBe(document.activeElement);
  });
});
