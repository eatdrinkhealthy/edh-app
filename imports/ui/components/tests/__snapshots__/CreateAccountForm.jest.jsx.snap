// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateAccountForm /> matches render snapshot - without form validation errors 1`] = `
<div
  className="pt2 pl4 pr4 pb4"
>
  <form
    className="measure center"
    onSubmit={[Function]}
  >
    <fieldset
      className="ba b--transparent ph0 mh0"
      id="sign_up"
    >
      <legend
        className="f4 fw6 ph0 mh0"
      >
        Create Account
      </legend>
      <div
        className="mt3"
      >
        <div>
          <label
            className="db fw6 lh-copy f6"
            htmlFor="username"
          >
            Username
          </label>
          <input
            autoFocus={true}
            className="pa2 input-reset ba bg-transparent hover-bg-black hover-white w-100"
            id="username"
            name="username"
            onBlur={[Function]}
            onChange={[Function]}
            onDragStart={[Function]}
            onDrop={[Function]}
            onFocus={[Function]}
            placeholder="Username"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        className="mt3"
      >
        <div>
          <label
            className="db fw6 lh-copy f6"
            htmlFor="email"
          >
            Email
          </label>
          <input
            autoFocus={undefined}
            className="pa2 input-reset ba bg-transparent hover-bg-black hover-white w-100"
            id="email"
            name="email"
            onBlur={[Function]}
            onChange={[Function]}
            onDragStart={[Function]}
            onDrop={[Function]}
            onFocus={[Function]}
            placeholder="Email"
            type="email"
            value=""
          />
        </div>
      </div>
      <div
        className="mv3"
      >
        <div>
          <label
            className="db fw6 lh-copy f6"
            htmlFor="password"
          >
            Password
          </label>
          <input
            autoFocus={undefined}
            className="pa2 input-reset ba bg-transparent hover-bg-black hover-white w-100"
            id="password"
            name="password"
            onBlur={[Function]}
            onChange={[Function]}
            onDragStart={[Function]}
            onDrop={[Function]}
            onFocus={[Function]}
            placeholder="Password"
            type="password"
            value=""
          />
        </div>
      </div>
      <div
        className="mv3"
      >
        <div>
          <label
            className="db fw6 lh-copy f6"
            htmlFor="confirmPassword"
          >
            Confirm Password
          </label>
          <input
            autoFocus={undefined}
            className="pa2 input-reset ba bg-transparent hover-bg-black hover-white w-100"
            id="confirmPassword"
            name="confirmPassword"
            onBlur={[Function]}
            onChange={[Function]}
            onDragStart={[Function]}
            onDrop={[Function]}
            onFocus={[Function]}
            placeholder="Confirm Password"
            type="password"
            value=""
          />
        </div>
      </div>
    </fieldset>
    <div>
      <input
        className="b ph3 pv2 input-reset ba b--black bg-transparent grow pointer f6 dib"
        type="submit"
        value="Create Account"
      />
    </div>
  </form>
</div>
`;
